File: ios/HTTPOverBluetooth/Sources/HTTPOverBluetooth/Common/HTTPResponse.swift
import Foundation

/// Represents an HTTP response received over BLE.
public struct HTTPResponse: Sendable {
    /// The HTTP status code
    public let statusCode: Int
    
    /// HTTP response headers
    public let headers: [String: String]
    
    /// Response body data
    public let body: Data?
    
    /// Whether this was an HTTPS response
    public let isHTTPS: Bool
    
    /// Whether the certificate was validated (for HTTPS)
    public let certificateValidated: Bool
    
    /// Creates a new HTTP response.
    /// - Parameters:
    ///   - statusCode: The HTTP status code
    ///   - headers: HTTP response headers
    ///   - body: Response body data
    ///   - isHTTPS: Whether this was an HTTPS response
    ///   - certificateValidated: Whether the certificate was validated
    public init(
        statusCode: Int,
        headers: [String: String] = [:],
        body: Data? = nil,
        isHTTPS: Bool = false,
        certificateValidated: Bool = false
    ) {
        self.statusCode = statusCode
        self.headers = headers
        self.body = body
        self.isHTTPS = isHTTPS
        self.certificateValidated = certificateValidated
    }
    
    /// Parses HTTP headers from Data.
    /// - Parameter data: The data containing headers
    /// - Returns: A dictionary of header key-value pairs
    public static func parseHeaders(from data: Data) -> [String: String] {
        guard let headerString = String(data: data, encoding: .utf8) else {
            return [:]
        }
        
        var headers: [String: String] = [:]
        let lines = headerString.components(separatedBy: "\r\n")
        
        for line in lines {
            if let colonIndex = line.firstIndex(of: ":") {
                let key = String(line[..<colonIndex]).trimmingCharacters(in: .whitespaces)
                let value = String(line[line.index(after: colonIndex)...]).trimmingCharacters(in: .whitespaces)
                headers[key] = value
            }
        }
        
        return headers
    }
    
    /// Parses HTTP status code from Data.
    /// The status code is encoded as a 16-bit unsigned integer (little-endian).
    /// - Parameter data: The data containing the status code
    /// - Returns: The HTTP status code
    public static func parseStatusCode(from data: Data) -> Int {
        guard data.count >= 2 else { return 0 }
        return Int(data[0]) | (Int(data[1]) << 8)
    }
    
    /// Serializes status code to Data for transmission.
    /// - Parameter statusCode: The HTTP status code
    /// - Returns: The status code as Data
    public static func serializeStatusCode(_ statusCode: Int) -> Data {
        var data = Data(count: 2)
        data[0] = UInt8(statusCode & 0xFF)
        data[1] = UInt8((statusCode >> 8) & 0xFF)
        return data
    }
}

File: android/http-over-bluetooth/src/main/java/com/example/httpoverble/common/HttpResponse.kt
package com.example.httpoverble.common

/**
 * Represents an HTTP response received over BLE.
 */
data class HttpResponse(
    val statusCode: Int,
    val headers: Map<String, String> = emptyMap(),
    val body: ByteArray? = null,
    val isHttps: Boolean = false,
    val certificateValidated: Boolean = false
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as HttpResponse

        if (statusCode != other.statusCode) return false
        if (headers != other.headers) return false
        if (body != null) {
            if (other.body == null) return false
            if (!body.contentEquals(other.body)) return false
        } else if (other.body != null) return false
        if (isHttps != other.isHttps) return false
        if (certificateValidated != other.certificateValidated) return false

        return true
    }

    override fun hashCode(): Int {
        var result = statusCode
        result = 31 * result + headers.hashCode()
        result = 31 * result + (body?.contentHashCode() ?: 0)
        result = 31 * result + isHttps.hashCode()
        result = 31 * result + certificateValidated.hashCode()
        return result
    }

    companion object {
        /**
         * Parses HTTP headers from a byte array.
         */
        fun parseHeaders(data: ByteArray): Map<String, String> {
            val headerString = String(data, Charsets.UTF_8)
            val headers = mutableMapOf<String, String>()
            headerString.split("\r\n").forEach { line ->
                val colonIndex = line.indexOf(':')
                if (colonIndex > 0) {
                    val key = line.substring(0, colonIndex).trim()
                    val value = line.substring(colonIndex + 1).trim()
                    headers[key] = value
                }
            }
            return headers
        }

        /**
         * Parses HTTP status code from a byte array.
         * The status code is encoded as a 16-bit unsigned integer (little-endian).
         */
        fun parseStatusCode(data: ByteArray): Int {
            if (data.size < 2) return 0
            return (data[0].toInt() and 0xFF) or ((data[1].toInt() and 0xFF) shl 8)
        }

        /**
         * Creates status code bytes for transmission.
         */
        fun serializeStatusCode(statusCode: Int): ByteArray {
            return byteArrayOf(
                (statusCode and 0xFF).toByte(),
                ((statusCode shr 8) and 0xFF).toByte()
            )
        }
    }
}
